packages/@aws-cdk/aws-codepipeline/lib/pipeline.ts

```diff
@@ -447,6 +444,36 @@ export class Pipeline extends PipelineBase {
     return !owner || owner === 'AWS';
   }
 
+  private getOtherStackIfActionIsCrossAccount(action: IAction): Stack | undefined {
+    const pipelineStack = Stack.of(this);
+
+    if (action.actionProperties.resource) {
+      const resourceStack = Stack.of(action.actionProperties.resource);
+      // check if resource is from a different account
+      if (pipelineStack.account !== resourceStack.account) {
+        return resourceStack;
+      }
+    } else if (action.actionProperties.account &&
```

Comment:
rewrite with early bailout style to simplify logic

Reply:
This isn't actually easy here either. If you change the current:

```ts
    if (action.actionProperties.resource) {
      // ...
    } else if (action.actionProperties.account &&
        pipelineStack.account !== action.actionProperties.account) {
      // ...
   }
```

to

```ts
    if (action.actionProperties.resource) {
      // ...
    }
     if (action.actionProperties.account &&
        pipelineStack.account !== action.actionProperties.account) {
      // ...
    }
```

that's actually a significant change in the logic - the first one ignores `account` completely if `resource` is set (which I believe is the correct behavior), while the other only does it if `resource` is from a different Stack.

Reply:
Obvsiouly I didn't intend for you to change the logic, but rather that you will retain the logic by splitting it up to smaller logical blocks... Let me try to propose:

```ts
  private getOtherStackIfActionIsCrossAccount(action: IAction): Stack | undefined {
    const pipelineStack = Stack.of(this);

    // if the action has a backing resource in a different account, return it's stack.
    if (action.actionProperties.resource) {
      const resourceStack = Stack.of(action.actionProperties.resource);

      // check if resource is from a different account
      if (pipelineStack.account !== resourceStack.account) {
        return resourceStack;
      }

      return undefined;
    }

    // an explicit account is specified at the action level
    if (action.actionProperties.account) {

      // we can't afford env-agnostic shit here.
      if (Token.isUnresolved(pipelineStack.account)) { throw new Error('...'); }
      if (Token.isUnresolved(action.actionProperties.account)) { throw new Error('...'); }

      // same account, done.
      if (action.actionProperties.account === pipelineStack.account) {
        return undefined;
      }

      return this.getCreateCrossAccountSupportStack(action.actionProperties.account);
    }

    // neither "resource" or "account" are set for the action, so we don't have a cross account thing going.
    return undefined;
  }

  private getCreateCrossAccountSupportStack(targetAccount: string): Stack {
    const cached = this._crossAccountSupport[targetAccount];
    if (cached) { return cached; }

    const app = this.requireApp();
    const stack = new Stack(app, `cross-account-support-stack-${targetAccount}`, {
      stackName: `${pipelineStack.stackName}-support-${targetAccount}`,
      env: {
        account: targetAccount,
        region: action.actionProperties.region ? action.actionProperties.region : pipelineStack.region,
      },
    });
    this._crossAccountSupport[targetAccount] = stack;

    return stack;
  }
```

Reply:
Refactored to reduce nesting. Let me know if you like this way better.